src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNVectorScoreScript.java

```diff
@@ -0,0 +1,166 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.script;
+
+import com.amazon.opendistroforelasticsearch.knn.index.util.KNNConstants;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.script.ScoreScript;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.UncheckedIOException;
+import java.util.Map;
+
+/**
+ * Vector score script used for adjusting the score based on similarity space
+ * on a per document basis.
+ *
+ */
+public class KNNVectorScoreScript extends ScoreScript {
+
+    private BinaryDocValues binaryDocValuesReader;
+    private final float[] queryVector;
+    private final String similaritySpace;
+    private float queryVectorSquaredMagnitude = -1;
+
+    /**
+     * This function called for each doc in the segment. We evaluate the score of the vector in the doc
+     *
+     * @param explanationHolder A helper to take in an explanation from a script and turn
+     *                          it into an {@link org.apache.lucene.search.Explanation}
+     * @return score of the vector to the query vector
+     */
+    @Override
+    public double execute(ScoreScript.ExplanationHolder explanationHolder) {
+        float score = Float.MIN_VALUE;
+        try {
+            float[] doc_vector;
+            BytesRef bytesref = binaryDocValuesReader.binaryValue();
+            // If there is no vector for the corresponding doc then it should be not considered for nearest
+            // neighbors.
+            if (bytesref == null) {
+                return Float.MIN_VALUE;
+            }
+            try (ByteArrayInputStream byteStream = new ByteArrayInputStream(bytesref.bytes, bytesref.offset, bytesref.length);
+                 ObjectInputStream objectStream = new ObjectInputStream(byteStream)) {
+                doc_vector = (float[]) objectStream.readObject();
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+
+            if(doc_vector.length != queryVector.length) {
+                throw new IllegalStateException("[KNN] query vector and field vector dimensions mismatch. " +
+                        "query vector: " + queryVector.length + ", stored vector: " + doc_vector.length);
+            }
+
+            if (KNNConstants.L2.equalsIgnoreCase(similaritySpace)) {
+                score = KNNScoringUtil.l2Squared(this.queryVector, doc_vector);
+                score = 1/(1 + score);
+            } else if (KNNConstants.COSINESIMIL.equalsIgnoreCase(similaritySpace)) {
+                // Scores cannot be negative so add +1 to the cosine score
+                score = 1 + KNNScoringUtil.cosinesimilOptimized(this.queryVector, doc_vector, this.queryVectorSquaredMagnitude);
```

Comment:
From https://en.wikipedia.org/wiki/Cosine_similarity,  seems others generally use `1 - cosine similarity`, any special logic here we use `1 + cosine similarity` ?

Reply:
Make sense. So the greater score means it's more relevant?

Reply:
Correct.